include/vkwrap/vkwrap/device.h

```diff
@@ -0,0 +1,407 @@
+
+    VulkanVersion m_version = VulkanVersion::e_version_1_0; // Version the physical device has to support
+    WeightFunction m_weight_function = []( auto&& ) -> int {
+        return 0;
+    };
+
+  public:
+    template <ranges::range Range> PhysicalDeviceSelector& withExtensions( Range&& extensions ) &
+    {
+        ranges::copy( std::forward<Range>( extensions ), ranges::back_inserter( m_extensions ) );
+        return *this;
+    }
+
+    PhysicalDeviceSelector& withPresent( vk::SurfaceKHR surface ) &
+    {
+        assert( surface && "Empty surface handle" );
+        m_surface = surface;
+        return *this;
+    } // withPresent
+
+    PhysicalDeviceSelector& withVersion( VulkanVersion version ) &
+    {
+        m_version = version;
+        return *this;
+    } // withVersion
+
+    PhysicalDeviceSelector& withWeight( WeightFunction function ) &
+    {
+        assert( function );
+        m_weight_function = function;
+        return *this;
+    } // withVersion
+
+    static constexpr auto k_default_types =
+        std::array{ vk::PhysicalDeviceType::eDiscreteGpu, vk::PhysicalDeviceType::eIntegratedGpu };
+
+    template <ranges::range Range = decltype( ( k_default_types ) )> // The second nested parenthesis here are crucial
+    PhysicalDeviceSelector& withTypes( Range&& types = k_default_types ) &
+    {
+        // clang-format off
+        auto zipped_view = ranges::views::zip_with( 
+            []( auto&& a, auto&& b ) { return std::pair{ a, b }; }, 
+            types, ranges::views::iota( 0 )
+        ); // clang-format on
+
+        m_types.clear();
+        ranges::copy( zipped_view, ranges::inserter( m_types, m_types.end() ) );
+        return *this;
+    } // withTypes
+
+  private:
+    auto calculateWeight( const PhysicalDeviceInfo& elem ) const -> std::pair<PhysicalDeviceInfo, int>
+    {
+        auto [ device, properties, id ] = elem;
+        constexpr int invalid = -1;
+
+        // Call custom function to skip doing unnecessary work if possible.
+        auto weight = m_weight_function( elem );
+
+        // Check basic requirements.
+        bool valid = ( weight >= 0 ) && device.supportsExtensions( m_extensions ).supports &&
+            m_types.contains( properties.deviceType );
+
+        if ( m_surface )
+        {
+            valid = valid && physicalDeviceSupportsPresent( device.get(), m_surface );
+        }
+
+        int cost = valid ? weight + m_types.at( properties.deviceType ) : invalid;
+        return std::pair{ elem, cost };
+    } // calculateWeight
+
+    using WeightMap = std::unordered_map<PhysicalDeviceInfo, int>;
+
+    WeightMap makeWeighted( const vk::Instance& instance ) const
+    {
+        const auto weight_calculation = [ & ]( auto&& elem ) {
+            return calculateWeight( elem );
+        };
+
+        auto all_devices = instance.enumeratePhysicalDevices();
+        WeightMap weight_map;
+
+        const auto get_properites = []( auto&& device ) {
```

Comment:
Why the lambda is generic, the *enumeratePhysicalDevices()* return a concrete type: *vk::PhysicalDevice*. Seems like over engineering.